test/differentials/composite.jl

```diff
@@ -65,6 +67,8 @@ end
         # Testing iterate via collect
         @test collect(Composite{Foo}(x=2.5)) == [2.5]
         @test collect(Composite{Tuple{Float64,}}(2.0)) == [2.0]
+        @test collect(Float64, Composite{Tuple{Float64,}}(@thunk 2.0^2)) == [4.0]
+        @test collect(Float64, Composite{Tuple{Float64,}}(a=(@thunk 2.0^2),)) == [4.0]
```

Comment:
We don't have tests for dictionary backed composites.
That is probably ok to continue to not have them here
for now.

THough I suspect iterating them will be weird, since they will iterate pairs and only the value, not the whole pair will need to be unthunked.
So perhaps it is worth making sure we do them right in this PR

Reply:
> THough I suspect iterating them will be weird, since they will iterate pairs and only the value, not the whole pair will need to be unthunked.

Agreed.
> We don't have tests for dictionary backed composites.
That is probably ok to continue to not have them here
for now.

I rather think that having the implementation of `Composite{<:Dict, <:Dict}` would be better here, since I think that approach of `iterate` might work better in general for `Dict` and `Tuple`'s, as the next state for a `Composite` of Tuple's is `state + 1` but may not so for `Dict`.

WDYT, shall I try that in this PR?

Reply:
We usually get `element, state = iterate(collection)` so, we would get the state for free by using  the `iterate` approach, but not from the `getindex` approach.